api/apps/api/src/modules/clone/export/adapters/node-manifest-file-service.adapter.ts

```diff
@@ -0,0 +1,175 @@
+import { Inject, Injectable } from '@nestjs/common';
+import { spawnSync } from 'child_process';
+import { createSign } from 'crypto';
+import { Either, left, right } from 'fp-ts/lib/Either';
+import { isLeft } from 'fp-ts/lib/These';
+import { Readable } from 'stream';
+import { FileDestination } from '../application/archive-creator.port';
+import {
+  integrityCheckFailed,
+  invalidSignature,
+  manifestFileGenerationError,
+  CloningSigningSecret,
+  ManifestFileService,
+  signatureFileGenerationError,
+} from '../application/manifest-file-service.port';
+import { ExportId } from '../domain';
+
+@Injectable()
+export class NodeManifestFileService implements ManifestFileService {
+  constructor(
+    private readonly fileRepository: CloningFilesRepository,
+    @Inject(CloningSigningSecret)
+    private readonly cloningSigningSecret: string,
+  ) {}
+
+  private async getManifestFileReadable(
+    exportFolder: string,
+  ): Promise<Either<typeof manifestFileGenerationError, Buffer>> {
+    const childProcess = spawnSync(
+      `find ${exportFolder} -type f -exec sha256sum {} \\;`,
+      { shell: true },
+    );
+
+    if (childProcess.error) {
+      return left(manifestFileGenerationError);
+    }
+
+    return right(childProcess.stdout);
+  }
+
+  performIntegrityCheck(
+    exportId: ExportId,
+  ): Either<typeof integrityCheckFailed, true> {
+    const exportFolder = this.fileRepository.getFilesFolderFor(exportId.value);
+
+    const childProcess = spawnSync(
+      `sha256sum -c ${exportFolder}/manifest.txt`,
+      { shell: true },
+    );
+
+    const stdout = childProcess.stdout.toString();
+    const stderr = childProcess.stderr.toString();
+    const error = stdout.includes('FAILED') || stderr.length > 0;
+
+    if (error) {
+      return left(integrityCheckFailed);
+    }
+
+    return right(true);
+  }
+
+  async verifyManifestFileSignature(
+    manifestFileUri: string,
+  ): Promise<Either<typeof invalidSignature, true>> {
+    try {
+      const manifestFile = await this.fileRepository.get(manifestFileUri);
+
+      if (isLeft(manifestFile)) {
+        throw manifestFile;
+      }
+
+      const manifestFileBuffer = await readableToBuffer(manifestFile.right);
+
+      const signer = createSign('RSA-SHA256');
+      signer.update(manifestFileBuffer);
+      const recalculatedSignature = signer.sign(
+        this.cloningSigningSecret,
+        'hex',
+      );
+
+      const signatureFileUri = manifestFileUri.replace(
+        manifestFileRelativePath,
+        signatureFileRelativePath,
+      );
+      const signatureFile = await this.fileRepository.get(signatureFileUri);
+      if (isLeft(signatureFile)) {
+        throw signatureFile;
+      }
+      const signatureFileBuffer = await readableToBuffer(signatureFile.right);
+      const uploadedSignature = signatureFileBuffer.toString();
+
+      if (recalculatedSignature !== uploadedSignature)
+        return left(invalidSignature);
```

Comment:
I think most of this could be done via something like this:
```suggestion
      const signatureFile = await this.fileRepository.get(signatureFileUri);
      if (isLeft(signatureFile)) {
        throw signatureFile;
      }
      const signatureFileBuffer = await readableToBuffer(signatureFile.right);
      const uploadedSignature = signatureFileBuffer.toString();

      const verifier = createVerify('RSA-SHA256');
      verifier.update(manifestFileBuffer);
      return verifier.verify(
        this.cloningSigningSecret,
        uploadedSignature,
        'hex',
      ) ? right(true) : left(invalidSignature);
```

Reply:
Yes, we will change it 👍